File: backend/services/scan_service.py
# backend/services/scan_service.py
from backend.services.firebase_admin import firestore_client
from typing import Dict, List
import uuid
from datetime import datetime

SCANS_COLLECTION = "scans"

class ScanService:
    def __init__(self):
        self.db = firestore_client

    def start_scan(self, target: str, user_id: str) -> Dict:
        scan_id = str(uuid.uuid4())
        scan_data = {
            "scan_id": scan_id,
            "user_id": user_id,
            "target": target,
            "status": "started",
            "progress": 0,
            "created_at": datetime.utcnow(),
            "finished_at": None,
            "findings": [],
        }
        self.db.collection(SCANS_COLLECTION).document(scan_id).set(scan_data)
        return scan_data

    def get_scan(self, scan_id: str) -> Dict:
        doc = self.db.collection(SCANS_COLLECTION).document(scan_id).get()
        if doc.exists:
            return doc.to_dict()
        raise Exception("Scan not found")

    def list_scans(self, user_id: str = None) -> List[Dict]:
        query = self.db.collection(SCANS_COLLECTION)
        if user_id:
            query = query.where("user_id", "==", user_id)
        return [doc.to_dict() for doc in query.stream()]

    def update_scan_status(self, scan_id: str, status: str, progress: int = None):
        update_data = {"status": status}
        if progress is not None:
            update_data["progress"] = progress
        if status == "finished":
            update_data["finished_at"] = datetime.utcnow()
        self.db.collection(SCANS_COLLECTION).document(scan_id).update(update_data)
        return self.get_scan(scan_id)

File: frontends/src/services/scanService.ts
// src/services/scanService.ts
import apiClient from "./client";

export interface Scan {
  scan_id: string;
  target: string;
  status: string;
  finished_at?: string | null;
}

// List all scans
export async function listScans(): Promise<Scan[]> {
  const res = await apiClient.get("/scan");
  return res.data ?? [];
}

// Start a new scan
export async function startScan(target: string): Promise<Scan> {
  const res = await apiClient.post("/scan/start", { target });
  return res.data;
}

// Get a single scan by ID
export async function getScan(id: string): Promise<Scan> {
  const res = await apiClient.get(`/scan/${id}`);
  return res.data;
}

// Cancel a running scan
export async function cancelScan(id: string): Promise<{ success: boolean }> {
  const res = await apiClient.post(`/scan/${id}/cancel`);
  return res.data;
}
